File: smartmerge/core/folder_compare.py
"""Folder/directory comparison logic."""

from pathlib import Path
from enum import Enum
from typing import NamedTuple, Optional
from .diff_engine import myers_opcodes


class ItemType(Enum):
    """Type of item in folder."""

    FILE = "file"
    FOLDER = "folder"


class ItemStatus(Enum):
    """Status of item relative to both sides."""

    IDENTICAL = "identical"
    MODIFIED = "modified"
    ADDED_LEFT = "added_left"
    ADDED_RIGHT = "added_right"
    DELETED_LEFT = "deleted_left"
    DELETED_RIGHT = "deleted_right"
    FOLDER_LEFT_ONLY = "folder_left_only"
    FOLDER_RIGHT_ONLY = "folder_right_only"


class FolderItem(NamedTuple):
    """Represents a file or folder in comparison."""

    name: str
    type: ItemType
    status: ItemStatus
    left_path: Optional[Path]
    right_path: Optional[Path]


def compare_folders(left_path: Path, right_path: Path) -> list[FolderItem]:
    """
    Compare two folders at root level (non-recursive).

    Returns list of FolderItem entries with status for each item.
    Items present only on one side will have None for the missing path.
    """
    if not left_path.is_dir() or not right_path.is_dir():
        return []

    # Get items in each folder
    left_items = {p.name: p for p in left_path.iterdir()}
    right_items = {p.name: p for p in right_path.iterdir()}

    all_names = set(left_items.keys()) | set(right_items.keys())
    results = []

    for name in sorted(all_names):
        left_p = left_items.get(name)
        right_p = right_items.get(name)

        # Both sides exist
        if left_p and right_p:
            if left_p.is_dir() and right_p.is_dir():
                # Both are folders - always identical (no recursive compare)
                status = ItemStatus.IDENTICAL
                item_type = ItemType.FOLDER
            elif left_p.is_file() and right_p.is_file():
                # Both are files - compare contents
                item_type = ItemType.FILE
                status = _compare_files(left_p, right_p)
            else:
                # One is file, one is folder - treat as modified
                status = ItemStatus.MODIFIED
                item_type = ItemType.FILE

            results.append(
                FolderItem(
                    name=name,
                    type=item_type,
                    status=status,
                    left_path=left_p,
                    right_path=right_p,
                )
            )

        # Only on left
        elif left_p:
            item_type = ItemType.FOLDER if left_p.is_dir() else ItemType.FILE
            status = (
                ItemStatus.FOLDER_LEFT_ONLY
                if left_p.is_dir()
                else ItemStatus.ADDED_LEFT
            )
            results.append(
                FolderItem(
                    name=name,
                    type=item_type,
                    status=status,
                    left_path=left_p,
                    right_path=None,
                )
            )

        # Only on right
        else:
            item_type = ItemType.FOLDER if right_p.is_dir() else ItemType.FILE
            status = (
                ItemStatus.FOLDER_RIGHT_ONLY
                if right_p.is_dir()
                else ItemStatus.ADDED_RIGHT
            )
            results.append(
                FolderItem(
                    name=name,
                    type=item_type,
                    status=status,
                    left_path=None,
                    right_path=right_p,
                )
            )

    return results


def _compare_files(left_path: Path, right_path: Path) -> ItemStatus:
    """Compare two files and return status."""
    try:
        with open(left_path, "r", encoding="utf-8", errors="ignore") as f:
            left_content = f.read()
    except (OSError, IOError):
        return ItemStatus.MODIFIED

    try:
        with open(right_path, "r", encoding="utf-8", errors="ignore") as f:
            right_content = f.read()
    except (OSError, IOError):
        return ItemStatus.MODIFIED

    if left_content == right_content:
        return ItemStatus.IDENTICAL

    # Use Myers diff to check if files differ
    opcodes = myers_opcodes(
        left_content.splitlines(keepends=False),
        right_content.splitlines(keepends=False),
    )

    return ItemStatus.IDENTICAL if not opcodes else ItemStatus.MODIFIED


def is_folder_navigable(item: FolderItem) -> bool:
    """
    Check if a folder item can be navigated into.

    Only folders that exist on both sides can be navigated.
    """
    return (
        item.type == ItemType.FOLDER
        and item.left_path is not None
        and item.right_path is not None
    )

File: smartmerge/src-tauri/src/folder_compare.rs
use std::collections::BTreeMap;
use std::fs;
use std::path::Path;

use crate::models::{FolderItem, ItemStatus, ItemType};

pub fn compare_folders(left_path: &str, right_path: &str) -> Result<Vec<FolderItem>, String> {
    let left_dir = Path::new(left_path);
    let right_dir = Path::new(right_path);

    if !left_dir.is_dir() || !right_dir.is_dir() {
        return Ok(Vec::new());
    }

    let left_entries = read_entries(left_dir)?;
    let right_entries = read_entries(right_dir)?;

    let mut all_names = BTreeMap::new();
    for name in left_entries.keys() {
        all_names.insert(name.clone(), ());
    }
    for name in right_entries.keys() {
        all_names.insert(name.clone(), ());
    }

    let mut results = Vec::new();

    for name in all_names.keys() {
        let left = left_entries.get(name);
        let right = right_entries.get(name);

        match (left, right) {
            (Some(left_pathbuf), Some(right_pathbuf)) => {
                let left_is_dir = left_pathbuf.is_dir();
                let right_is_dir = right_pathbuf.is_dir();
                if left_is_dir && right_is_dir {
                    results.push(FolderItem {
                        name: name.clone(),
                        item_type: ItemType::Folder,
                        status: ItemStatus::Identical,
                        left_path: Some(left_pathbuf.to_string_lossy().to_string()),
                        right_path: Some(right_pathbuf.to_string_lossy().to_string()),
                    });
                } else if !left_is_dir && !right_is_dir {
                    let status = compare_files(left_pathbuf, right_pathbuf);
                    results.push(FolderItem {
                        name: name.clone(),
                        item_type: ItemType::File,
                        status,
                        left_path: Some(left_pathbuf.to_string_lossy().to_string()),
                        right_path: Some(right_pathbuf.to_string_lossy().to_string()),
                    });
                } else {
                    results.push(FolderItem {
                        name: name.clone(),
                        item_type: ItemType::File,
                        status: ItemStatus::Modified,
                        left_path: Some(left_pathbuf.to_string_lossy().to_string()),
                        right_path: Some(right_pathbuf.to_string_lossy().to_string()),
                    });
                }
            }
            (Some(left_pathbuf), None) => {
                let item_type = if left_pathbuf.is_dir() {
                    ItemType::Folder
                } else {
                    ItemType::File
                };
                let status = if left_pathbuf.is_dir() {
                    ItemStatus::FolderLeftOnly
                } else {
                    ItemStatus::AddedLeft
                };
                results.push(FolderItem {
                    name: name.clone(),
                    item_type,
                    status,
                    left_path: Some(left_pathbuf.to_string_lossy().to_string()),
                    right_path: None,
                });
            }
            (None, Some(right_pathbuf)) => {
                let item_type = if right_pathbuf.is_dir() {
                    ItemType::Folder
                } else {
                    ItemType::File
                };
                let status = if right_pathbuf.is_dir() {
                    ItemStatus::FolderRightOnly
                } else {
                    ItemStatus::AddedRight
                };
                results.push(FolderItem {
                    name: name.clone(),
                    item_type,
                    status,
                    left_path: None,
                    right_path: Some(right_pathbuf.to_string_lossy().to_string()),
                });
            }
            (None, None) => {}
        }
    }

    Ok(results)
}

pub fn is_folder_navigable(item: &FolderItem) -> bool {
    matches!(item.item_type, ItemType::Folder) && item.left_path.is_some() && item.right_path.is_some()
}

fn read_entries(path: &Path) -> Result<BTreeMap<String, std::path::PathBuf>, String> {
    let mut entries = BTreeMap::new();
    let dir_entries = fs::read_dir(path).map_err(|err| err.to_string())?;
    for entry in dir_entries {
        let entry = entry.map_err(|err| err.to_string())?;
        let path = entry.path();
        if let Some(name) = path.file_name().and_then(|n| n.to_str()) {
            entries.insert(name.to_string(), path);
        }
    }
    Ok(entries)
}

fn compare_files(left: &Path, right: &Path) -> ItemStatus {
    let left_content = match fs::read_to_string(left) {
        Ok(content) => content,
        Err(_) => return ItemStatus::Modified,
    };
    let right_content = match fs::read_to_string(right) {
        Ok(content) => content,
        Err(_) => return ItemStatus::Modified,
    };

    if left_content == right_content {
        ItemStatus::Identical
    } else {
        ItemStatus::Modified
    }
}
